20221012-add-type-restrictions-to-dsl-syntax.md

```diff
@@ -0,0 +1,140 @@
+ 2. Suppose A is an authorization model with type document and relation reader on document. We may have a type restriction on A that takes (document, reader) to the set {(user, \_), (group, member)}, meaning that the only users that may be readers of a document are users with type user, or a userset of group members.
+
+### Syntax
+
+Let A be an authorization model with set of types T and set of relations R, and recall that R* is the set R with the addition of an "empty" relation. We will write an element (t, r) of TxR* as t#r. The special case (t, _) is simply written as t. If the type restriction takes (t, r) to a set {(t1, r1), (t2, r2)}, then in our model we will write this as
+```
+type t
+    relations
+        define r: [t1#r1, t2#r2] ...
+```
+If the type restriction takes (t, r) to the empty set we will simply write nothing in our model.
+
+### Types with no relations
+
+It now becomes useful to allow types with no relations. For example, suppose we want to define a `viewer` relation and retrict the typeswe may want to specify a `user` type which is the type for all concrete users. This will look like
+```
+type user
+```
+
+### The dropping of "self"
+
+Notice that the addition of type restrictions means we may drop `self` from the DSL. This is because if a pair (type, relation) has a non-empty set of type restrictions, then the assumption is that (type, relation) is a pair that allows a direct relation to type. Thus, previously if we defined a relation as
+```
+define viewer as self
+```
+we can now write this as
+```
+define viewer: [user]
+```
+
+The dropping of "self" can be extended to rewrite rules as well. If a rewrite contains `self` then it may be dropped as in the following examples:
+- `define viewer as self or viewer from parent` -> `define viewer: [user] or viewer from parent`
+- `define viewer as self and viewer from parent` -> `define viewer: [user] and viewer from parent`
+- `define viewer as self but not banned` -> `define viewer: [user] but not banned`
+
+## Example
+
+Let's look at an example. Consider the model without type restrictions and with self:
+```
+type group
+    relations
+        define parent as self
+        define member as self
+        define guest as self or member
+        define can_view as guest
+```
+Suppose we want the following restrictions:
+1. parents of groups are groups
+2. members of groups are employees or group members
+3. guests of groups are users, employees, or group members
+
+We can write the model with these type restriction as
+```
+model
+    schema 1.1
+
+type user
+
+type employee
+
+type group
+    relations
+        define parent: [group]
+        define member: [employee, group#member]
+        define guest: [user, employee, group#member] or member
+        define can_view as guest
+```
+
```

Comment:
Please add wildcard as part of type restriction in the RFCs

Reply:
Also, I thought we are standardizing to use `:` instead of `as`

Reply:
I'll add wildcard. Thank you!

I hadn't heard about the dropping of `as`. Let me check. Thanks!